src/nnet3bin/nnet3-egs-augment-image.cc

```diff
@@ -63,77 +64,73 @@ struct ImageAugmentationConfig {
   }
 };
 
-
-// Flips the image horizontally.
-void HorizontalFlip(MatrixBase<BaseFloat> *image) {
-  int32 num_rows = image->NumRows();
-  Vector<BaseFloat> temp(image->NumCols());
-  for (int32 r = 0; r < num_rows / 2; r++) {
-    SubVector<BaseFloat> row_a(*image, r), row_b(*image,
-                                                 num_rows - r - 1);
-    temp.CopyFromVec(row_a);
-    row_a.CopyFromVec(row_b);
-    row_b.CopyFromVec(temp);
-  }
-}
-
- // Shifts the image horizontally by 'horizontal_shift' (+ve == to the right).
-void HorizontalShift(int32 horizontal_shift,
-                     MatrixBase<BaseFloat> *image) {
-  int32 num_rows = image->NumRows();
-  for (int32 r = 0; r < num_rows; r++) {
-    int32 current_r;
-    // +ve == to the right, do shifting from right to left; otherwise, from left to right
-    if (horizontal_shift > 0) {
-      current_r = r;
-    } else {  // do it in the reverse order.
-      current_r = num_rows - 1 - r;
-    }
-
-    int32 origin_r = current_r - horizontal_shift;
-    if (origin_r < 0)
-      origin_r = 0;
-    if (origin_r >= num_rows)
-      origin_r = num_rows - 1;
-
-    SubVector<BaseFloat> current_row(*image, current_r),
-        origin_row(*image, origin_r);
-    current_row.CopyFromVec(origin_row);
-  }
-}
-
-/* Shifts the image vertically by 'vertical_shift' (+ve == to the top).*/
-void VerticalShift(int32 vertical_shift,
-                   int32 num_channels,
-                   MatrixBase<BaseFloat> *image) {
+void ApplyAffineTransform(MatrixBase<BaseFloat> &transform,
+                          int32 num_channels,
+                          MatrixBase<BaseFloat> *image) {
   int32 num_rows = image->NumRows(),
-      num_cols = image->NumCols(), height = num_cols / num_channels;
+        num_cols = image->NumCols(),
+        height = num_cols / num_channels;
   KALDI_ASSERT(num_cols % num_channels == 0);
+  Matrix<BaseFloat> original_image(*image);
   for (int32 r = 0; r < num_rows; r++) {
-    BaseFloat *this_row = image->RowData(r);
     for (int32 c = 0; c < height; c++) {
-      int32 current_index;
-      // +ve == to the top, do shifting from top to bottom; otherwise, bottom to top
-      if (vertical_shift > 0) {
-        current_index = height - 1 - c;
-      } else {
-        current_index = c;
+      BaseFloat r_old = transform(0, 0) * r +
+                                          transform(0, 1) * c + transform(0, 2);
+      BaseFloat c_old = transform(1, 0) * r +
+                                          transform(1, 1) * c + transform(1, 2);
+
+      // we are going to do bilinear interpolation between 4 closest points
+      // to the point [r_old, c_old]. We have: r1  <=  r_old  <=  r2
+      //                                       c1  <=  c_old  <=  c2
+
+      int32 r1 = static_cast<int32>(floor(r_old));
+      int32 c1 = static_cast<int32>(floor(c_old));
+      int32 r2 = r1 + 1;
+      int32 c2 = c1 + 1;
+
```

Comment:
I think it would be cleaner if you obtained the interpolation factors at this stage, before modifying r1, r2, c1 and c2.
E.g.
`BaseFloat weight_11 = ..., weight_12 = ...`
Then you don't have to modify r_old and  c_old, which is confusing.
Changing it like this would also be more efficient.